src/lib/transition_frontier/extensions/best_tip_diff.ml

```diff
@@ -77,16 +124,17 @@ module T = struct
         ~f:
           (fun ( ({ new_commands; removed_commands; reorg_best_tip = _ } as acc)
                , should_broadcast ) -> function
-            | E (Best_tip_changed new_best_tip, old_best_tip_hash) ->
-                let new_best_tip_breadcrumb =
-                  Full_frontier.find_exn frontier new_best_tip
-                in
-                let old_best_tip =
-                  (*FIXME #4404*)
-                  Full_frontier.find_exn frontier old_best_tip_hash
-                in
+            | E (Best_tip_changed new_best_tip_hash, old_best_tip_hash) ->
                 let added_to_best_tip_path, removed_from_best_tip_path =
-                  get_path_diff t frontier new_best_tip_breadcrumb old_best_tip
+                  load_paths ~new_best_tip_hash ~old_best_tip_hash t frontier
+                  |> Result.map_error ~f:(fun (hash, message) ->
+                         Error.of_string
+                         @@ sprintf
+                              "failed to retrieve hash %s from frontier in \
+                               best_tip_diff extension: %s"
+                              (State_hash.to_base58_check hash)
+                              message )
+                  |> Or_error.ok_exn
```

Comment:
I pushed exception throwing to this single place in `best_tip_diff` extension handling, and code that is called from here is handled via `Result.t` monad.